crates/codegen/language/definition/src/model/utils/mod.rs

```diff
@@ -1,5 +1,78 @@
+use crate::model::{self, Item};
+use semver::Version;
+use std::collections::BTreeSet;
+
 mod identifier;
 mod version_specifier;
 
 pub use identifier::*;
 pub use version_specifier::*;
+
+/// Collects all versions that changes the language grammar in a breaking way.
+pub fn collect_breaking_versions(lang: &model::Language) -> BTreeSet<Version> {
+    let first = lang.versions.first().cloned().unwrap();
+    let mut res = BTreeSet::from_iter([first]);
+
+    fn add_spec(res: &mut BTreeSet<Version>, spec: &Option<model::VersionSpecifier>) {
+        let Some(spec) = spec else {
+            return;
+        };
+
+        match spec.clone() {
+            model::VersionSpecifier::Never => (),
+            model::VersionSpecifier::From { from } => {
+                res.insert(from);
+            }
+            model::VersionSpecifier::Till { till } => {
+                res.insert(till);
+            }
+            model::VersionSpecifier::Range { from, till } => {
+                res.insert(from);
+                res.insert(till);
+            }
+        }
+    }
+
+    for item in lang
+        .sections
+        .iter()
+        .flat_map(|s| &s.topics)
+        .flat_map(|t| &t.items)
+    {
+        match item.as_ref() {
+            Item::Struct { item } => {
+                add_spec(&mut res, &item.enabled);
+                for field in &item.fields {
+                    match field.1 {
+                        model::Field::Required { .. } => (),
+                        model::Field::Optional { enabled, .. } => add_spec(&mut res, enabled),
+                    }
+                }
+            }
+            Item::Enum { item } => {
+                add_spec(&mut res, &item.enabled);
+                for variant in &item.variants {
+                    add_spec(&mut res, &variant.enabled);
+                }
+            }
+            Item::Repeated { item } => add_spec(&mut res, &item.enabled),
+            Item::Separated { item } => add_spec(&mut res, &item.enabled),
+            Item::Precedence { item } => add_spec(&mut res, &item.enabled),
```

Comment:
I think we also need to collect specifiers inside individual operators under `Precedence`.

Reply:
Done in d18e93a61465bbe231337d126b4c99e913c35afc

Reply:
How about `primary_expressions`?

Reply:
Ah, I somehow missed that these are versioned as well. Good catch! Done in https://github.com/NomicFoundation/slang/pull/684/commits/f3fa3a7ddea2a521a4b9fd61f24101b73aaf991a

Reply:
:finnadie: Pushed in 862efc3d308fe88d52169163f2d4538c56bdc49c